iceoryx_utils/include/iceoryx_utils/internal/cxx/storable_function.hpp

```diff
@@ -0,0 +1,162 @@
+#ifndef IOX_UTILS_STORABLE_FUNCTION_HPP
+#define IOX_UTILS_STORABLE_FUNCTION_HPP
+
+#include "iceoryx_utils/cxx/function_ref.hpp"
+#include "iceoryx_utils/cxx/type_traits.hpp"
+#include "iceoryx_utils/internal/cxx/static_storage.hpp"
+
+#include <iostream>
+#include <type_traits>
+#include <utility>
+
+namespace iox
+{
+namespace cxx
+{
+template <typename ReturnType, typename... Args>
+using signature = ReturnType(Args...);
+
+template <typename StorageType, typename T>
+class storable_function;
+
+template <typename StorageType, typename ReturnType, typename... Args>
+class storable_function<StorageType, signature<ReturnType, Args...>>
+{
+  private:
+    // note that this vtable or a similar approach with virtual is needed to ensure we perform the correct
+    // operation with the underlying (erased) type
+    // this means storable_function cannot be used where pointers become invalid, e.g. across process boundaries
+    struct vtable
+    {
+        // exposing those is intentional, it is an internal structure
+        void (*copyFunction)(const storable_function& src, storable_function& dest){nullptr};
+        void (*moveFunction)(storable_function& src, storable_function& dest){nullptr};
+        void (*destroyFunction)(storable_function& f){nullptr};
+
+        vtable() = default;
+        vtable(const vtable& other) = default;
+        vtable& operator=(const vtable& other) = default;
+        vtable(vtable&& other) = default;
+        vtable& operator=(vtable&& other) = default;
+
+        void copy(const storable_function& src, storable_function& dest)
+        {
+            if (copyFunction)
+            {
+                copyFunction(src, dest);
+            }
+        }
+
+        void move(storable_function& src, storable_function& dest)
+        {
+            if (moveFunction)
+            {
+                moveFunction(src, dest);
+            }
+        }
+
+        void destroy(storable_function& f)
+        {
+            if (destroyFunction)
+            {
+                destroyFunction(f);
+            }
+        }
+    };
+
+    vtable m_vtable;
+
+    /// @note  in general we cannot know the alignment of the type we want to store at construction time
+    StorageType m_storage;
+    void* m_storedObj{nullptr};
+    function_ref<signature<ReturnType, Args...>> m_function;
+
+  public:
+    using signature_t = signature<ReturnType, Args...>;
+
+    storable_function() = default;
+
+    /// @brief construct from functor (including lambda)
+    template <typename Functor,
+              typename = typename std::enable_if<std::is_class<Functor>::value
+                                                     && is_invocable_r<ReturnType, Functor, Args...>::value,
+                                                 void>::type>
+    storable_function(const Functor& functor) noexcept;
+
+    /// @brief construct from function pointer (including static functions)
+    storable_function(ReturnType (*function)(Args...)) noexcept;
```

Comment:
Because the cases are handled differently internally and too generic constructors with universal references cause all kinds of issues, it is especially easy to overlook a type category that can be passed into them and will not actually work.